packages/wrangler/src/init.ts

```diff
@@ -620,3 +688,191 @@ async function findPath(
+
+			switch (binding.type) {
+				case "plain_text":
+					{
+						configObj["vars"] = {
+							...(configObj?.var ?? {}),
+							name: binding.name,
+							text: binding.text,
+						};
+					}
+					break;
+				case "json":
+					{
+						configObj["vars"] = {
+							...(configObj?.var ?? {}),
+							name: binding.name,
+							json: binding.json,
+						};
+					}
+					break;
+				case "kv_namespace":
+					{
+						configObj["kv_namespaces"] = [
+							...(configObj?.["kv_namespace"] ?? []),
+							{ id: binding.namespace_id, binding: binding.name },
+						];
+					}
+					break;
+				case "durable_object_namespace":
+					{
+						configObj["durable_objects"] = {
+							bindings: [
+								...(configObj?.durable_objects?.bindings ?? []),
+								{
+									name: binding.name,
+									class_name: binding.class_name,
+									script_name: binding.script_name,
+									environment: binding.environment,
+								},
+							],
+						};
+					}
+					break;
+				case "r2_bucket":
+					{
+						configObj["r2_buckets"] = [
+							...(configObj?.r2_buckets ?? []),
+							{ binding: binding.name, bucket_name: binding.bucket_name },
+						];
+					}
+					break;
+				case "service":
+					{
+						configObj["service"] = [
+							...(configObj?.service ?? []),
+							{
+								binding: binding.name,
+								service: binding.service,
+								environment: binding.environment,
+							},
+						];
+					}
+					break;
+				case "namespace":
+					{
+						configObj["namespaces"] = [
+							...(configObj?.namespaces ?? []),
+							{ binding: binding.name, namespace: binding.namespace },
+						];
+					}
+					break;
+				case "logfwdr":
+					{
+						configObj["logfwdr"] = {
+							// TODO: Messaging about adding schema file path
+							schema: "",
+							bindings: [
+								...(configObj?.logfwdr?.bindings ?? []),
+								{ name: binding.name, destination: binding.destination },
+							],
+						};
+					}
+					break;
+				case "wasm_module":
+					{
+						configObj["wasm_modules"] = {
+							...(configObj?.wasm_modules ?? {}),
+							[binding.name]: binding.part,
+						};
+					}
+					break;
+				case "text_blob":
+					{
+						configObj["text_blobs"] = {
+							...(configObj?.text_blobs ?? {}),
+							[binding.name]: binding.part,
+						};
+					}
+					break;
+				case "data_blob":
+					{
+						configObj["data_blobs"] = {
+							...(configObj?.data_blobs ?? {}),
+							[binding.name]: binding.part,
+						};
+					}
+					break;
+				default: {
+					// If we don't know what the type is, its an unsafe binding
+					// eslint-disable-next-line @typescript-eslint/no-explicit-any
+					if (!(binding as any)?.type) break;
+					configObj["unsafe"] = {
+						bindings: [...(configObj?.unsafe?.bindings ?? []), binding],
+					};
+				}
+			}
+
+			return configObj;
+			// eslint-disable-next-line @typescript-eslint/no-explicit-any
+		}, {} as Record<string, any>);
```

Comment:
Is there any way this could be typed better? Since we know all the binding types, it seems this type could be constructed without `any` for better type safety (and avoiding the pluralisation bugs above)

Reply:
Not really, the structure of the config is by definition unknown/any due to it potentially being any permutation of the config in Dashboard. I am open to ideas of how to get TypeScript to do it for us though, I just didn't want to end up with complex generics or manual typing that was incomplete.